3rdparty/openexr/CMakeLists.txt

```diff
@@ -76,6 +76,8 @@ set(OPENEXR_PACKAGE_STRING "\"OpenEXR ${OPENEXR_VERSION}\"" )
 
 configure_file("${CMAKE_CURRENT_SOURCE_DIR}/IlmBaseConfig.h.cmakein"
                "${CMAKE_CURRENT_BINARY_DIR}/IlmBaseConfig.h" @ONLY)
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/IlmThreadConfig.h.cmakein"
```

Comment:
See above

```
set(ILMBASE_VERSION_MAJOR "2")
set(ILMBASE_VERSION_MINOR "3")
set(ILMBASE_VERSION_PATCH "0")
```

These versions must be updated.

Reply:
Updated to 3.1.0.

For the record: I cloned the updated version from [here](https://github.com/AcademySoftwareFoundation/openexr/tree/4cc01e52c1cd5219340ac068ee6a47825b095c30). Had to do minor changes to avoid compiler warnings and such.

Reply:
> Had to do minor changes to avoid compiler warnings and such.

We usually don't modify 3rdparty source code.
We prefer to suppress warnings in CMakeLists.txt, see `ocv_warnings_disable()` calls.
We keep only "must have" patched which should be migrated during upgrade.

Right upgrade flow is the following:
- extract "patch" for modified code of legacy version
- upgrade source code to new version
- re-apply and adapt extracted patch.

---

> 3.1.0

Current latest release is `v3.1.3`: https://github.com/AcademySoftwareFoundation/openexr/releases
There are several bugfixed landed. Probably including some fixed CVEs from here: #21326

Reply:
Ok. I took the version from `master`, and it's from Dec 2021, whereas `v3.1.3` is from Oct 2021. I guess they just don't update the patch version number in `master`.

But I can try to revert to `v3.1.3`, if that's what you prefer?

Reply:
There are too much changes against master (100+ files): https://github.com/AcademySoftwareFoundation/openexr/compare/v3.1.3...master

if we declare 3.1.3 then we should use 3.1.3 code.
Please fetch code from `v3.1.3` tag.

Reply:
> We usually don't modify 3rdparty source code.

I think the most relevant change I had to do was [this](https://github.com/reunanen/opencv/commit/c38ebcc672ab2f5202dc168f6149125b05edf35b). It seems to me that the OpenEXR code assumes C++17 here. No problem in my project, but some of the OpenCV CI builds were complaining about this.

Reply:
Ok, this simple change is fine, especially as dedicated commit (OpenCV 4.x should work with C++11 compiler)

However, no idea about C++17:
- https://en.cppreference.com/w/cpp/numeric/math/fabs
- https://en.cppreference.com/w/cpp/numeric/math/abs

Reply:
Actually that one (`abs`/`fabs`) wasn't needed after all.

In the end of the day, I was able to avoid all other changes to Imath and OpenEXR code, except for these two, which I have left as separate commits:

- https://github.com/opencv/opencv/pull/21324/commits/b30a7d733e30eccd54843a2503fe950c9ccbc604 : compiler error about ambiguous symbol (namespace issue)
  - fixed by specifying the namespace explicitly
- https://github.com/opencv/opencv/pull/21324/commits/6ff64710ec1653d04873f657b610b34dea978f85 : warning about deprecated attribute
  - fixed by updating the attribute names